goa v2 gRPC custom errors validate on server side using wrong type

The code generated for custom error validation on server side uses the wrong type (or the wrong type is code generated):
`$GOPATH/src/foo/design/api.go`
```
package design

import (
	. "goa.design/goa/dsl"
)

var _ = Service("Foo", func() {
	Method("foobar", func() {
		Result(Bar)
		Error("foobar_error", FooBarError)
		GRPC(func() {
			Response(CodeOK)
			Response("foobar_error", CodeFailedPrecondition)
		})
	})
})

var Bar = ResultType("application/vnd.foo.bar", func() {
	TypeName("Bar")
	Attributes(func() {
		Field(1, "name", String)
		Required("name")
	})
	View("default", func() {
		Attribute("name")
	})
})

var FooBarError = ResultType("application/vnd.foo.foobar_error", func() {
	TypeName("FooBarError")

	Attributes(func() {
		Field(1, "details")
		Field(2, "id", String)
		Field(3, "name", String, func() {
			Meta("struct:error:name")
			Enum("foobar_error")
		})
		Required("details", "id", "name")
	})
})
```
The generated code does not compile:
```
$ goa gen foo/design
gen/foo/client.go
gen/foo/endpoints.go
gen/foo/service.go
gen/foo/views/view.go
gen/grpc/cli/foo/cli.go
gen/grpc/foo/client/cli.go
gen/grpc/foo/client/client.go
gen/grpc/foo/client/encode_decode.go
gen/grpc/foo/client/types.go
gen/grpc/foo/pb/foo.proto
gen/grpc/foo/server/encode_decode.go
gen/grpc/foo/server/server.go
gen/grpc/foo/server/types.go
$
$ go build foo/gen/grpc/foo/server
# foo/gen/grpc/foo/server
gen/grpc/foo/server/types.go:40:35: undefined: foopb.FooBarError
```
The generated `foo/gen/grpc/foo/server/types.go` file is using the wrong type for validation (or else the type name was generated incorrectly):
```
// Code generated by goa v2.0.0-wip, DO NOT EDIT.
//
// Foo gRPC server types
//
// Command:
// $ goa gen foo/design

package server

import (
	foo "foo/gen/foo"
	fooviews "foo/gen/foo/views"
	foopb "foo/gen/grpc/foo/pb"

	goa "goa.design/goa"
)

// NewFoobarResponse builds the gRPC response type from the result of the
// "foobar" endpoint of the "Foo" service.
func NewFoobarResponse(result *fooviews.BarView) *foopb.FoobarResponse {
	message := &foopb.FoobarResponse{}
	if result.Name != nil {
		message.Name = *result.Name
	}
	return message
}

// NewFoobarFoobarErrorError builds the gRPC error response type from the error
// of the "foobar" endpoint of the "Foo" service.
func NewFoobarFoobarErrorError(er *foo.FooBarError) *foopb.FoobarFoobarErrorError {
	message := &foopb.FoobarFoobarErrorError{
		Details: er.Details,
		Id:      er.ID,
		Name:    er.Name,
	}
	return message
}

// ValidateFooBarError runs the validations defined on FooBarError.
func ValidateFooBarError(message *foopb.FooBarError) (err error) {
        ^^^ message type should be *foopb.FoobarFoobarErrorError or else
        ^^^ the type name should have been generated as foopb.FooBarError,
        ^^^ which sounds better.
	if !(message.Name == "foobar_error") {
		err = goa.MergeErrors(err, goa.InvalidEnumValueError("message.name", message.Name, []interface{}{"foobar_error"}))
	}
	return
}
```

--- a/grpc/codegen/client.go
+++ b/grpc/codegen/client.go
@@ -211,6 +211,11 @@ func (c *{{ .ClientStruct }}) {{ .Method.VarName }}() goa.Endpoint {
 			{{- range .Errors }}
 				{{- if .Response.ClientConvert }}
 					case {{ .Response.ClientConvert.SrcRef }}:
+						{{- if .Response.ClientConvert.Validation }}
+							if err := {{ .Response.ClientConvert.Validation.Name }}(message); err != nil {
+								return nil, err
+							}
+						{{- end }}
 						return nil, {{ .Response.ClientConvert.Init.Name }}({{ range .Response.ClientConvert.Init.Args }}{{ .Name }}, {{ end }})
 				{{- end }}
 			{{- end }}

--- a/grpc/codegen/client_types.go
+++ b/grpc/codegen/client_types.go
@@ -89,7 +89,7 @@ func clientType(genpkg string, svc *expr.GRPCServiceExpr, seen map[string]struct
 			})
 		}
 		for _, data := range sd.validations {
-			if data.Request {
+			if data.Kind == "server" {
 				continue
 			}
 			sections = append(sections, &codegen.SectionTemplate{

--- a/grpc/codegen/client_types.go
+++ b/grpc/codegen/client_types.go
@@ -89,6 +89,9 @@ func clientType(genpkg string, svc *expr.GRPCServiceExpr, seen map[string]struct
 			})
 		}
 		for _, data := range sd.validations {
+			if data.Request {
+				continue
+			}
 			sections = append(sections, &codegen.SectionTemplate{
 				Name:   "client-validate",
 				Source: validateT,

--- a/grpc/codegen/client_types.go
+++ b/grpc/codegen/client_types.go
@@ -89,7 +89,7 @@ func clientType(genpkg string, svc *expr.GRPCServiceExpr, seen map[string]struct
 			})
 		}
 		for _, data := range sd.validations {
-			if data.Kind == "server" {
+			if data.Kind == validateServer {
 				continue
 			}
 			sections = append(sections, &codegen.SectionTemplate{

--- a/grpc/codegen/client_types_test.go
+++ b/grpc/codegen/client_types_test.go
@@ -17,6 +17,7 @@ func TestClientTypeFiles(t *testing.T) {
 	}{
 		{"payload-with-nested-types", testdata.PayloadWithNestedTypesDSL, testdata.PayloadWithNestedTypesClientTypeCode},
 		{"result-collection", testdata.ResultWithCollectionDSL, testdata.ResultWithCollectionClientTypeCode},
+		{"with-errors", testdata.UnaryRPCWithErrorsDSL, testdata.WithErrorsClientTypeCode},
 	}
 	for _, c := range cases {
 		t.Run(c.Name, func(t *testing.T) {

--- a/grpc/codegen/server_types.go
+++ b/grpc/codegen/server_types.go
@@ -89,6 +89,9 @@ func serverType(genpkg string, svc *expr.GRPCServiceExpr, seen map[string]struct
 			})
 		}
 		for _, data := range sd.validations {
+			if !data.Request {
+				continue
+			}
 			sections = append(sections, &codegen.SectionTemplate{
 				Name:   "server-validate",
 				Source: validateT,

--- a/grpc/codegen/server_types.go
+++ b/grpc/codegen/server_types.go
@@ -89,7 +89,7 @@ func serverType(genpkg string, svc *expr.GRPCServiceExpr, seen map[string]struct
 			})
 		}
 		for _, data := range sd.validations {
-			if data.Kind == "client" {
+			if data.Kind == validateClient {
 				continue
 			}
 			sections = append(sections, &codegen.SectionTemplate{

--- a/grpc/codegen/server_types.go
+++ b/grpc/codegen/server_types.go
@@ -89,7 +89,7 @@ func serverType(genpkg string, svc *expr.GRPCServiceExpr, seen map[string]struct
 			})
 		}
 		for _, data := range sd.validations {
-			if !data.Request {
+			if data.Kind == "client" {
 				continue
 			}
 			sections = append(sections, &codegen.SectionTemplate{

--- a/grpc/codegen/server_types_test.go
+++ b/grpc/codegen/server_types_test.go
@@ -17,6 +17,7 @@ func TestServerTypeFiles(t *testing.T) {
 	}{
 		{"payload-with-nested-types", testdata.PayloadWithNestedTypesDSL, testdata.PayloadWithNestedTypesServerTypeCode},
 		{"result-collection", testdata.ResultWithCollectionDSL, testdata.ResultWithCollectionServerTypeCode},
+		{"with-errors", testdata.UnaryRPCWithErrorsDSL, testdata.WithErrorsServerTypeCode},
 	}
 	for _, c := range cases {
 		t.Run(c.Name, func(t *testing.T) {

--- a/grpc/codegen/service_data.go
+++ b/grpc/codegen/service_data.go
@@ -246,6 +246,8 @@ type (
 		SrcName string
 		// SrcRef is the fully qualified reference to the type being validated.
 		SrcRef string
+		// Request if true indicates that the validation is for request messages.
+		Request bool
 	}
 
 	// InitData contains the data required to render a constructor.
@@ -639,7 +641,10 @@ func collectMessages(at *expr.AttributeExpr, sd *ServiceData, seen map[string]st
 // addValidation adds a validation function (if any) for the given user type
 // and recurses through the user type adding other validation functions
 // (if any).
-func addValidation(att *expr.AttributeExpr, sd *ServiceData) *ValidationData {
+//
+// req if true indicates that the validation is generated for validating
+// request (server-side) messages.
+func addValidation(att *expr.AttributeExpr, sd *ServiceData, req bool) *ValidationData {
 	ut, ok := att.Type.(expr.UserType)
 	if !ok {
 		return nil
@@ -659,16 +664,17 @@ func addValidation(att *expr.AttributeExpr, sd *ServiceData) *ValidationData {
 		}
 	}
 	ctx := protoBufTypeContext("", sd.Scope)
-	if def := codegen.RecursiveValidationCode(att, ctx, false, "message"); def != "" {
+	if def := codegen.RecursiveValidationCode(att, ctx, true, "message"); def != "" {
 		v := &ValidationData{
 			Name:    "Validate" + name,
 			Def:     def,
 			ArgName: "message",
 			SrcName: name,
 			SrcRef:  ref,
+			Request: req,
 		}
 		sd.validations = append(sd.validations, v)
-		collectValidations(att, ctx, true, sd)
+		collectValidations(att, ctx, req, sd)
 		return v
 	}
 	return nil
@@ -676,6 +682,9 @@ func addValidation(att *expr.AttributeExpr, sd *ServiceData) *ValidationData {
 
 // collectValidations recurses through the attribute and collects the
 // validation functions.
+//
+// req if true indicates that the validations are generated for validating
+// request messages.
 func collectValidations(att *expr.AttributeExpr, ctx *codegen.AttributeContext, req bool, sd *ServiceData, seen ...map[string]struct{}) {
 	var s map[string]struct{}
 	if len(seen) > 0 {
@@ -697,10 +706,11 @@ func collectValidations(att *expr.AttributeExpr, ctx *codegen.AttributeContext, 
 		s[name] = struct{}{}
 		sd.validations = append(sd.validations, &ValidationData{
 			Name:    "Validate" + name,
-			Def:     codegen.RecursiveValidationCode(att, ctx, req, "message"),
+			Def:     codegen.RecursiveValidationCode(att, ctx, true, "message"),
 			ArgName: "message",
 			SrcName: name,
 			SrcRef:  protoBufGoFullTypeRef(att, sd.PkgName, sd.Scope),
+			Request: req,
 		})
 		att := dt.Attribute()
 		if rt, ok := dt.(*expr.ResultTypeExpr); ok {
@@ -712,13 +722,13 @@ func collectValidations(att *expr.AttributeExpr, ctx *codegen.AttributeContext, 
 		collectValidations(att, ctx, req, sd, seen...)
 	case *expr.Object:
 		for _, nat := range *dt {
-			collectValidations(nat.Attribute, ctx, att.IsRequired(nat.Name), sd, seen...)
+			collectValidations(nat.Attribute, ctx, req, sd, seen...)
 		}
 	case *expr.Array:
-		collectValidations(dt.ElemType, ctx, true, sd, seen...)
+		collectValidations(dt.ElemType, ctx, req, sd, seen...)
 	case *expr.Map:
-		collectValidations(dt.KeyType, ctx, true, sd, seen...)
-		collectValidations(dt.ElemType, ctx, true, sd, seen...)
+		collectValidations(dt.KeyType, ctx, req, sd, seen...)
+		collectValidations(dt.ElemType, ctx, req, sd, seen...)
 	}
 }
 
@@ -774,7 +784,7 @@ func buildRequestConvertData(request, payload *expr.AttributeExpr, md []*Metadat
 			TgtName:    svc.Scope.GoFullTypeName(payload, svcCtx.Pkg),
 			TgtRef:     svc.Scope.GoFullTypeRef(payload, svcCtx.Pkg),
 			Init:       data,
-			Validation: addValidation(request, sd),
+			Validation: addValidation(request, sd, true),
 		}
 	}
 
@@ -872,7 +882,7 @@ func buildResponseConvertData(response, result *expr.AttributeExpr, svcCtx *code
 		TgtName:    svcCtx.Scope.Name(result, svcCtx.Pkg),
 		TgtRef:     svcCtx.Scope.Ref(result, svcCtx.Pkg),
 		Init:       data,
-		Validation: addValidation(response, sd),
+		Validation: addValidation(response, sd, false),
 	}
 }
 
@@ -1018,7 +1028,7 @@ func buildErrorConvertData(ge *expr.GRPCErrorExpr, e *expr.GRPCEndpointExpr, sd 
 		TgtName:    svcCtx.Scope.Name(ge.ErrorExpr.AttributeExpr, svcCtx.Pkg),
 		TgtRef:     svcCtx.Scope.Ref(ge.ErrorExpr.AttributeExpr, svcCtx.Pkg),
 		Init:       data,
-		Validation: addValidation(ge.ErrorExpr.AttributeExpr, sd),
+		Validation: addValidation(ge.Response.Message, sd, false),
 	}
 }
 
@@ -1077,7 +1087,7 @@ func buildStreamData(e *expr.GRPCEndpointExpr, sd *ServiceData, svr bool) *Strea
 					TgtName:    svcCtx.Scope.Name(e.MethodExpr.StreamingPayload, svcCtx.Pkg),
 					TgtRef:     recvRef,
 					Init:       buildInitData(e.StreamingRequest, e.MethodExpr.StreamingPayload, "v", "spayload", svcCtx, false, sd),
-					Validation: addValidation(e.StreamingRequest, sd),
+					Validation: addValidation(e.StreamingRequest, sd, true),
 				}
 			}
 			mustClose = md.ServerStream.MustClose
@@ -1106,7 +1116,7 @@ func buildStreamData(e *expr.GRPCEndpointExpr, sd *ServiceData, svr bool) *Strea
 					TgtName:    resCtx.Scope.Name(result, resCtx.Pkg),
 					TgtRef:     resCtx.Scope.Ref(result, resCtx.Pkg),
 					Init:       buildInitData(e.Response.Message, result, "v", resVar, resCtx, false, sd),
-					Validation: addValidation(e.Response.Message, sd),
+					Validation: addValidation(e.Response.Message, sd, false),
 				}
 			}
 			mustClose = md.ClientStream.MustClose

--- a/grpc/codegen/service_data.go
+++ b/grpc/codegen/service_data.go
@@ -249,7 +249,7 @@ type (
 		// Kind indicates that the validation is for request (server-side),
 		// response (client-side), or both (server and client side) messages.
 		// It is used to generate validation code in the server and client packages.
-		Kind string
+		Kind validateKind
 	}
 
 	// InitData contains the data required to render a constructor.
@@ -338,12 +338,28 @@ type (
 		// for the stream.
 		MustClose bool
 	}
+
+	// validateKind is a type to determine where the validation code is generated
+	// (server, client, or both)
+	validateKind int
 )
 
 const (
 	// pbPkgName is the directory name where the .proto file is generated and
 	// compiled.
 	pbPkgName = "pb"
+)
+
+const (
+	// validateServer generates the validation code for request messages in the
+	// server package.
+	validateServer validateKind = iota + 1
+	// validateClient generates the validation code for response messages in the
+	// client package.
+	validateClient
+	// validateBoth generates the validation code in both server and client
+	// packages.
+	validateBoth
 )
 
 // Get retrieves the transport data for the service with the given name
@@ -653,14 +669,14 @@ func addValidation(att *expr.AttributeExpr, sd *ServiceData, req bool) *Validati
 	}
 	name := protoBufGoTypeName(att, sd.Scope)
 	ref := protoBufGoFullTypeRef(att, sd.PkgName, sd.Scope)
-	kind := "client"
+	kind := validateClient
 	if req {
-		kind = "server"
+		kind = validateServer
 	}
 	for _, n := range sd.validations {
 		if n.SrcName == name {
 			if n.Kind != kind {
-				n.Kind = "both"
+				n.Kind = validateBoth
 			}
 			return n
 		}
@@ -707,14 +723,14 @@ func collectValidations(att *expr.AttributeExpr, ctx *codegen.AttributeContext, 
 		if _, ok := s[name]; ok {
 			return
 		}
-		kind := "client"
+		kind := validateClient
 		if req {
-			kind = "server"
+			kind = validateServer
 		}
 		for _, n := range sd.validations {
 			if n.SrcName == name {
 				if n.Kind != kind {
-					n.Kind = "both"
+					n.Kind = validateBoth
 				}
 				return
 			}

--- a/grpc/codegen/service_data.go
+++ b/grpc/codegen/service_data.go
@@ -246,8 +246,10 @@ type (
 		SrcName string
 		// SrcRef is the fully qualified reference to the type being validated.
 		SrcRef string
-		// Request if true indicates that the validation is for request messages.
-		Request bool
+		// Kind indicates that the validation is for request (server-side),
+		// response (client-side), or both (server and client side) messages.
+		// It is used to generate validation code in the server and client packages.
+		Kind string
 	}
 
 	// InitData contains the data required to render a constructor.
@@ -651,8 +653,15 @@ func addValidation(att *expr.AttributeExpr, sd *ServiceData, req bool) *Validati
 	}
 	name := protoBufGoTypeName(att, sd.Scope)
 	ref := protoBufGoFullTypeRef(att, sd.PkgName, sd.Scope)
+	kind := "client"
+	if req {
+		kind = "server"
+	}
 	for _, n := range sd.validations {
 		if n.SrcName == name {
+			if n.Kind != kind {
+				n.Kind = "both"
+			}
 			return n
 		}
 	}
@@ -671,7 +680,7 @@ func addValidation(att *expr.AttributeExpr, sd *ServiceData, req bool) *Validati
 			ArgName: "message",
 			SrcName: name,
 			SrcRef:  ref,
-			Request: req,
+			Kind:    kind,
 		}
 		sd.validations = append(sd.validations, v)
 		collectValidations(att, ctx, req, sd)
@@ -698,8 +707,15 @@ func collectValidations(att *expr.AttributeExpr, ctx *codegen.AttributeContext, 
 		if _, ok := s[name]; ok {
 			return
 		}
+		kind := "client"
+		if req {
+			kind = "server"
+		}
 		for _, n := range sd.validations {
 			if n.SrcName == name {
+				if n.Kind != kind {
+					n.Kind = "both"
+				}
 				return
 			}
 		}
@@ -710,7 +726,7 @@ func collectValidations(att *expr.AttributeExpr, ctx *codegen.AttributeContext, 
 			ArgName: "message",
 			SrcName: name,
 			SrcRef:  protoBufGoFullTypeRef(att, sd.PkgName, sd.Scope),
-			Request: req,
+			Kind:    kind,
 		})
 		att := dt.Attribute()
 		if rt, ok := dt.(*expr.ResultTypeExpr); ok {

--- a/grpc/codegen/testdata/client_endpoint_init_code.go
+++ b/grpc/codegen/testdata/client_endpoint_init_code.go
@@ -80,8 +80,14 @@ func (c *Client) MethodUnaryRPCWithErrors() goa.Endpoint {
 			resp := goagrpc.DecodeError(err)
 			switch message := resp.(type) {
 			case *service_unaryrpc_with_errorspb.MethodUnaryRPCWithErrorsInternalError:
+				if err := ValidateMethodUnaryRPCWithErrorsInternalError(message); err != nil {
+					return nil, err
+				}
 				return nil, NewMethodUnaryRPCWithErrorsInternalError(message)
 			case *service_unaryrpc_with_errorspb.MethodUnaryRPCWithErrorsBadRequestError:
+				if err := ValidateMethodUnaryRPCWithErrorsBadRequestError(message); err != nil {
+					return nil, err
+				}
 				return nil, NewMethodUnaryRPCWithErrorsBadRequestError(message)
 			case *service_unaryrpc_with_errorspb.MethodUnaryRPCWithErrorsCustomErrorError:
 				return nil, NewMethodUnaryRPCWithErrorsCustomErrorError(message)

--- a/grpc/codegen/testdata/client_type_code.go
+++ b/grpc/codegen/testdata/client_type_code.go
@@ -14,35 +14,6 @@ func NewMethodPayloadWithNestedTypesRequest(payload *servicepayloadwithnestedtyp
 	return message
 }
 
-// ValidateMethodPayloadWithNestedTypesRequest runs the validations defined on
-// MethodPayloadWithNestedTypesRequest.
-func ValidateMethodPayloadWithNestedTypesRequest(message *service_payload_with_nested_typespb.MethodPayloadWithNestedTypesRequest) (err error) {
-	if message.AParams != nil {
-		if err2 := ValidateAParams(message.AParams); err2 != nil {
-			err = goa.MergeErrors(err, err2)
-		}
-	}
-	return
-}
-
-// ValidateAParams runs the validations defined on AParams.
-func ValidateAParams(message *service_payload_with_nested_typespb.AParams) (err error) {
-
-	return
-}
-
-// ValidateArrayOfString runs the validations defined on ArrayOfString.
-func ValidateArrayOfString(message *service_payload_with_nested_typespb.ArrayOfString) (err error) {
-
-	return
-}
-
-// ValidateBParams runs the validations defined on BParams.
-func ValidateBParams(message *service_payload_with_nested_typespb.BParams) (err error) {
-
-	return
-}
-
 // protobufServicePayloadWithNestedTypespbAParamsToServicepayloadwithnestedtypesAParams
 // builds a value of type *servicepayloadwithnestedtypes.AParams from a value
 // of type *service_payload_with_nested_typespb.AParams.
@@ -192,5 +163,78 @@ func protobufServiceResultWithCollectionpbResultTToServiceresultwithcollectionRe
 	}
 
 	return res
+}
+`
+
+const WithErrorsClientTypeCode = `// NewMethodUnaryRPCWithErrorsRequest builds the gRPC request type from the
+// payload of the "MethodUnaryRPCWithErrors" endpoint of the
+// "ServiceUnaryRPCWithErrors" service.
+func NewMethodUnaryRPCWithErrorsRequest(payload string) *service_unaryrpc_with_errorspb.MethodUnaryRPCWithErrorsRequest {
+	message := &service_unaryrpc_with_errorspb.MethodUnaryRPCWithErrorsRequest{}
+	message.Field = payload
+	return message
+}
+
+// NewMethodUnaryRPCWithErrorsResult builds the result type of the
+// "MethodUnaryRPCWithErrors" endpoint of the "ServiceUnaryRPCWithErrors"
+// service from the gRPC response type.
+func NewMethodUnaryRPCWithErrorsResult(message *service_unaryrpc_with_errorspb.MethodUnaryRPCWithErrorsResponse) string {
+	result := message.Field
+	return result
+}
+
+// NewMethodUnaryRPCWithErrorsInternalError builds the error type of the
+// "MethodUnaryRPCWithErrors" endpoint of the "ServiceUnaryRPCWithErrors"
+// service from the gRPC error response type.
+func NewMethodUnaryRPCWithErrorsInternalError(message *service_unaryrpc_with_errorspb.MethodUnaryRPCWithErrorsInternalError) *serviceunaryrpcwitherrors.AnotherError {
+	er := &serviceunaryrpcwitherrors.AnotherError{
+		Name: message.Name,
+	}
+	if message.Description != "" {
+		er.Description = &message.Description
+	}
+	return er
+}
+
+// NewMethodUnaryRPCWithErrorsBadRequestError builds the error type of the
+// "MethodUnaryRPCWithErrors" endpoint of the "ServiceUnaryRPCWithErrors"
+// service from the gRPC error response type.
+func NewMethodUnaryRPCWithErrorsBadRequestError(message *service_unaryrpc_with_errorspb.MethodUnaryRPCWithErrorsBadRequestError) *serviceunaryrpcwitherrors.AnotherError {
+	er := &serviceunaryrpcwitherrors.AnotherError{
+		Name: message.Name,
+	}
+	if message.Description != "" {
+		er.Description = &message.Description
+	}
+	return er
+}
+
+// NewMethodUnaryRPCWithErrorsCustomErrorError builds the error type of the
+// "MethodUnaryRPCWithErrors" endpoint of the "ServiceUnaryRPCWithErrors"
+// service from the gRPC error response type.
+func NewMethodUnaryRPCWithErrorsCustomErrorError(message *service_unaryrpc_with_errorspb.MethodUnaryRPCWithErrorsCustomErrorError) *serviceunaryrpcwitherrors.ErrorType {
+	er := &serviceunaryrpcwitherrors.ErrorType{}
+	if message.A != "" {
+		er.A = &message.A
+	}
+	return er
+}
+
+// ValidateMethodUnaryRPCWithErrorsInternalError runs the validations defined
+// on MethodUnaryRPCWithErrorsInternalError.
+func ValidateMethodUnaryRPCWithErrorsInternalError(message *service_unaryrpc_with_errorspb.MethodUnaryRPCWithErrorsInternalError) (err error) {
+	if !(message.Name == "this" || message.Name == "that") {
+		err = goa.MergeErrors(err, goa.InvalidEnumValueError("message.name", message.Name, []interface{}{"this", "that"}))
+	}
+	return
+}
+
+// ValidateMethodUnaryRPCWithErrorsBadRequestError runs the validations defined
+// on MethodUnaryRPCWithErrorsBadRequestError.
+func ValidateMethodUnaryRPCWithErrorsBadRequestError(message *service_unaryrpc_with_errorspb.MethodUnaryRPCWithErrorsBadRequestError) (err error) {
+	if !(message.Name == "this" || message.Name == "that") {
+		err = goa.MergeErrors(err, goa.InvalidEnumValueError("message.name", message.Name, []interface{}{"this", "that"}))
+	}
+	return
 }
 `

--- a/grpc/codegen/testdata/dsls.go
+++ b/grpc/codegen/testdata/dsls.go
@@ -61,6 +61,7 @@ var UnaryRPCWithErrorsDSL = func() {
 		Attributes(func() {
 			Attribute("name", String, func() {
 				Meta("struct:error:name")
+				Enum("this", "that")
 			})
 			Attribute("description", String)
 			Required("name")

--- a/grpc/codegen/testdata/server_type_code.go
+++ b/grpc/codegen/testdata/server_type_code.go
@@ -194,3 +194,58 @@ func protobufServiceResultWithCollectionpbResultTToServiceresultwithcollectionRe
 	return res
 }
 `
+
+const WithErrorsServerTypeCode = `// NewMethodUnaryRPCWithErrorsPayload builds the payload of the
+// "MethodUnaryRPCWithErrors" endpoint of the "ServiceUnaryRPCWithErrors"
+// service from the gRPC request type.
+func NewMethodUnaryRPCWithErrorsPayload(message *service_unaryrpc_with_errorspb.MethodUnaryRPCWithErrorsRequest) string {
+	v := message.Field
+	return v
+}
+
+// NewMethodUnaryRPCWithErrorsResponse builds the gRPC response type from the
+// result of the "MethodUnaryRPCWithErrors" endpoint of the
+// "ServiceUnaryRPCWithErrors" service.
+func NewMethodUnaryRPCWithErrorsResponse(result string) *service_unaryrpc_with_errorspb.MethodUnaryRPCWithErrorsResponse {
+	message := &service_unaryrpc_with_errorspb.MethodUnaryRPCWithErrorsResponse{}
+	message.Field = result
+	return message
+}
+
+// NewMethodUnaryRPCWithErrorsInternalError builds the gRPC error response type
+// from the error of the "MethodUnaryRPCWithErrors" endpoint of the
+// "ServiceUnaryRPCWithErrors" service.
+func NewMethodUnaryRPCWithErrorsInternalError(er *serviceunaryrpcwitherrors.AnotherError) *service_unaryrpc_with_errorspb.MethodUnaryRPCWithErrorsInternalError {
+	message := &service_unaryrpc_with_errorspb.MethodUnaryRPCWithErrorsInternalError{
+		Name: er.Name,
+	}
+	if er.Description != nil {
+		message.Description = *er.Description
+	}
+	return message
+}
+
+// NewMethodUnaryRPCWithErrorsBadRequestError builds the gRPC error response
+// type from the error of the "MethodUnaryRPCWithErrors" endpoint of the
+// "ServiceUnaryRPCWithErrors" service.
+func NewMethodUnaryRPCWithErrorsBadRequestError(er *serviceunaryrpcwitherrors.AnotherError) *service_unaryrpc_with_errorspb.MethodUnaryRPCWithErrorsBadRequestError {
+	message := &service_unaryrpc_with_errorspb.MethodUnaryRPCWithErrorsBadRequestError{
+		Name: er.Name,
+	}
+	if er.Description != nil {
+		message.Description = *er.Description
+	}
+	return message
+}
+
+// NewMethodUnaryRPCWithErrorsCustomErrorError builds the gRPC error response
+// type from the error of the "MethodUnaryRPCWithErrors" endpoint of the
+// "ServiceUnaryRPCWithErrors" service.
+func NewMethodUnaryRPCWithErrorsCustomErrorError(er *serviceunaryrpcwitherrors.ErrorType) *service_unaryrpc_with_errorspb.MethodUnaryRPCWithErrorsCustomErrorError {
+	message := &service_unaryrpc_with_errorspb.MethodUnaryRPCWithErrorsCustomErrorError{}
+	if er.A != nil {
+		message.A = *er.A
+	}
+	return message
+}
+`

